### ISSUE TITLE
[1.8+?, Maybe 1.7.10 too?] Issue with AE2 Meteorite World Gen Support 

### ISSUE THREAD

**Issue by MJRLegends** [micdoodle8/Galacticraft#3105]
Was doing some testing of the MC 1.12.2 version of Galacticraft and noticed that the AE2 Meteorite World Gen at https://github.com/micdoodle8/Galacticraft/blob/MC1.12/src/main/java/micdoodle8/mods/galacticraft/core/TransformerHooks.java#L260 doesnt seem to be working.

I added -29 (Mars) to the meteoriteDimensionWhitelist list in the AE2 config see
```
    I:meteoriteDimensionWhitelist <
        -29
        0
     >
```
But they didnt seem to spawn/generate on Mars

Seems to be broken in MC 1.10.2 too, but for MC 1.10.2 it showed "Whitelisting AE2 meteorites worldgen on planets." in the log. Maybe we might need to check all versions for this....

**Comment by radfast** [comment]
Thanks, I'll prioritise this as we want worldgen to be good.

**Comment by MJRLegends** [comment]
No problem

**Comment by MJRLegends** [comment]
@radfast Also whats your view on adding support for other mods, for example the RecurrentComplex is a world gen spawning mod that you can add custom structures but this mod doesnt work with Galacticraft or Galacticraft addons. Is this is something we can look at on a large scale for integration with lots of other mods world gen

**Comment by radfast** [comment]
In Galacticraft the AE2 worldgen enablement is correctly working in 1.12.2.

If it's working in 1.12.2 then it should also be working in earlier MC versions, but if there's a specific one where you think it is not working, please say.

To test whether it is working on Galacticraft's side:
1.  Launch game, load a world
2.  Take the player to the Moon or any other planet  (space stations or Overworld won't work)
3.  If necessary walk the player to a new ungenerated spot on that world
4.  Check the console Game Output in the launcher for this line:
     [Galacticraft]: Whitelisting AE2 meteorites worldgen on planets.

You will only see that line in console **after** worldgen on a Galacticraft planet is initiated.  So you won't see it if your player stays on the Overworld, or if you are on a planet but you stay in an already-generated area.   Also, you will only see it once per Minecraft launch.

If you see that line, then the worldgen code in AE2 is being called on **all** Galacticraft planets, it doesn't matter which dimension so long as it is a ```IGalacticraftWorldProvider```.

If you are not seeing the expected worldgen in game (e.g. meteors) then the issue is probably on AE2's side.  I have had a quick look at their code and  I can see them doing a config check [here](https://github.com/AppliedEnergistics/Applied-Energistics-2/blob/rv5-1.12/src/main/java/appeng/worldgen/MeteoriteWorldGen.java#L45) and [this code](https://github.com/AppliedEnergistics/Applied-Energistics-2/blob/rv6-1.12/src/main/java/appeng/core/features/registries/WorldGenRegistry.java#L86-L113) **looks** to be OK but I have not tried debugging it.

**Comment by MJRLegends** [comment]
Did you see one spawn though?

**Comment by radfast** [comment]
No, none spawned.

I can see AE2 code is running OK on Mars in my game because in the world folder, ```AE2\spawndata```, there are NBT files being created with names ```-29.......```
This shows that their whitelisting code is working correctly, too.

------

OK I think I found the problem.  (They have a hard-coded list of blocks AE2 Meteorites can spawn on.)[https://github.com/AppliedEnergistics/Applied-Energistics-2/blob/rv5-1.12/src/main/java/appeng/worldgen/MeteoritePlacer.java#L80-L94]

Obviously, this hard-coded list does not include any Galacticraft blocks :(

**Comment by MJRLegends** [comment]
Yeah i probably just missed the console output on 1.12.2 but on 1.10.2 i saw the console mesaage but no actual spawning of one, also did you see my comment i made 13 hours ago

**Comment by radfast** [comment]
Nothing I can do about it.  We cannot even hack their list of blocks using Java reflection, because their code generates a **new instance** of MeteoritePlacer each time the worldgen is called (maybe even one for each chunk), and each instance creates its own hard-coded block list.

**Comment by radfast** [comment]
>  Also whats your view on adding support for other mods, for example the RecurrentComplex is a world gen spawning mod that you can add custom structures but this mod doesnt work with Galacticraft or Galacticraft addons. Is this is something we can look at on a large scale for integration with lots of other mods world gen

Maybe.  I'd need to add a new config file for this, which would need to contain the specific class names to be whitelisted.  And it can be generalised only for the "easy" ones - Thaumcraft is an example of one which is not an "easy" one.

There is an existing config option to enable ALL other mods worldgen, but you shouldn't use it because you'll get things like TConstruct Slime Islands and IC2 Rubber Trees on the moon - maybe even flowers from other mods.

**Comment by MJRLegends** [comment]
Yeah, even if we somehow could get RecurrentComplex mod support working players could probably use that mod to spawn ae2 meteors from coping one and making a custom structure from it

**Comment by radfast** [comment]
Give me a link to that mod's worldgen source code please.

**Comment by MJRLegends** [comment]
Here is the repo https://github.com/Ivorforce/RecurrentComplex but not sure where the actually code is for generation as only on my phone atm

**Comment by BrisingrAerowing** [comment]
Generation stuff is [in this folder](https://github.com/Ivorforce/RecurrentComplex/tree/master/src/main/java/ivorius/reccomplex/world/gen/feature).

Having Recurrent Complex support would be amazing for a high-tech pack I want to make (where Galacticraft will play a core role in progression).

**Comment by radfast** [comment]
Thanks @BrisingrAerowing for the link

I've now done the groundwork for this in https://github.com/micdoodle8/Galacticraft/commit/dfc2db10ed50dafeb5cf222455d613a301511847

Further progress will be a few days as I'm busy this week ahead.  I'll leave this open for now.

**Comment by radfast** [comment]
I should say: **I think I've found a way to workaround the AE2 hard-coded meteorites problem**

**Comment by MJRLegends** [comment]
Nice great work :D
